### PR TITLE
feat(plugin-chart-echarts): add x and y label support for 9 charts

### PR DESCRIPTION
**Pull request by MaxHuiYYDS**
fix #16512

🏆 Enhancements
This PR supports two features:
1 User can add X/Y axis label in below chart:
time series line chart,
time series area chart,
time series bar chat,
time series scatter plot,
mix time series,
time series chart,
time series smooth line,
time series stepped line,
box plot.
2 Changed previews default top Y axis position to dynamic left/top choice.
related to:
https://github.com/apache/superset/issues/16512

After:

https://user-images.githubusercontent.com/51693396/132269154-32f41add-c2ba-4764-9eee-20bfcbe877be.mov

### CODE REVIEW

**Comment by MaxHuiYYDS** [comment]
@junlincc  @stephenLYZ  please help me review this PR, :)

File: plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts [hunk 1]
```diff
@@ -64,6 +64,97 @@ export default {
       expanded: true,
       controlSetRows: [
         ['color_scheme'],
+        [
+          {
```
Comment (zhaoyongjie) [inline plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts:68]: Could we define these controls to `superset-ui/chart-controls`? (We should reuse same control)

File: plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts [hunk 2]
```diff
@@ -64,6 +64,97 @@ export default {
...
+            name: 'x_axis_label',
```
Comment (zhaoyongjie) [inline plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts:69]: Could we define these controls to `superset-ui/chart-controls`? (We should reuse same control)

File: plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts [hunk 3]
```diff
@@ -64,6 +64,97 @@ export default {
...
+              label: t('X Axis Title'),
+              renderTrigger: true,
+              default: '',
+              description: t('Changing this control takes effect instantly'),
+            },
+          },
+        ],
+        [
+          {
+            name: 'xAxisLabelBottomMargin',
```
Comment (zhaoyongjie) [inline plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts:81]: Same before.

File: plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts [hunk 4]
```diff
@@ -64,6 +64,97 @@ export default {
...
+              ],
+              default: '',
+              renderTrigger: true,
+              description: t('Changing this control takes effect instantly'),
+            },
+          },
+        ],
+        [
+          {
+            name: 'y_axis_label',
```
Comment (zhaoyongjie) [inline plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts:105]: Same before.

File: plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts [hunk 5]
```diff
@@ -64,6 +64,97 @@ export default {
...
+              label: t('Y Axis Title'),
+              renderTrigger: true,
+              default: '',
+              description: t('Changing this control takes effect instantly'),
+            },
+          },
+        ],
+        [
+          {
+            name: 'yAxisLabelPosition',
```
Comment (zhaoyongjie) [inline plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts:117]: Same before

File: plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts [hunk 6]
```diff
@@ -64,6 +64,97 @@ export default {
...
+              ],
+              default: '',
+              renderTrigger: true,
+              description: t('Changing this control takes effect instantly'),
+            },
+          },
+        ],
+        [
+          {
+            name: 'yAxisLabelMargin',
```
Comment (zhaoyongjie) [inline plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts:135]: Same before

File: plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts (hunk 2 above)
Comment (MaxHuiYYDS) [reply plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts]: Done

File: plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts (hunk 3 above)
Comment (MaxHuiYYDS) [reply plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts]: Done

File: plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts (hunk 4 above)
Comment (MaxHuiYYDS) [reply plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts]: Done

File: plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts (hunk 5 above)
Comment (MaxHuiYYDS) [reply plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts]: Done

File: plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts (hunk 6 above)
Comment (MaxHuiYYDS) [reply plugins/plugin-chart-echarts/src/BoxPlot/controlPanel.ts]: Done

File: packages/superset-ui-chart-controls/src/shared-controls/index.tsx [hunk 7]
```diff
@@ -495,6 +540,11 @@ const sharedControls = {
   adhoc_filters: enableExploreDnd ? dnd_adhoc_filters : adhoc_filters,
   color_scheme,
   label_colors,
+  x_axis_title,
```
Comment (zhaoyongjie) [inline packages/superset-ui-chart-controls/src/shared-controls/index.tsx:543]: Could we define a `section` to share all of these, please refer to: https://github.com/apache-superset/superset-ui/blob/master/packages/superset-ui-chart-controls/src/sections/forecastInterval.tsx

Comment (MaxHuiYYDS) [reply packages/superset-ui-chart-controls/src/shared-controls/index.tsx]: Done

File: packages/superset-ui-chart-controls/src/sections/chartTitle.tsx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+      {
+        name: 'x_axis_title_margin',
+        config: {
+          type: 'SelectControl',
+          freeForm: true,
+          clearable: false,
+          label: t('X AXIS TITLE BOTTOM MARGIN'),
+          renderTrigger: true,
+          default: TITLE_MARGIN_OPTIONS[0][0],
+          choices: TITLE_MARGIN_OPTIONS,
```
Comment (zhaoyongjie) [inline packages/superset-ui-chart-controls/src/sections/chartTitle.tsx:51]: define options like below:
```javascript
export const TITLE_MARGIN_OPTIONS: number[] = [
  15,
  30,
  50,
  75,
  100,
  125,
  150,
  200,
]
```

Use this way
```javascript
....
default: TITLE_MARGIN_OPTIONS[0],
choices: formatSelectOptions(TITLE_MARGIN_OPTIONS),
....
```

File: packages/superset-ui-chart-controls/src/shared-controls/index.tsx [hunk 9]
```diff
@@ -462,7 +462,6 @@ const label_colors: SharedControlConfig<'ColorMapControl'> = {
     colorScheme,
   }),
 };
-
```
Comment (zhaoyongjie) [inline packages/superset-ui-chart-controls/src/shared-controls/index.tsx:465]: don't change unrelated files.

File: packages/superset-ui-chart-controls/src/sections/chartTitle.tsx [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import React from 'react';
+import { t } from '@superset-ui/core';
+import { ControlPanelSectionConfig } from '../types';
+import { TITLE_MARGIN_OPTIONS, TITLE_POSITION_OPTIONS } from '../utils';
+
+export const titleControls: ControlPanelSectionConfig = {
+  label: t('Chart Title'),
```
Comment (zhaoyongjie) [inline packages/superset-ui-chart-controls/src/sections/chartTitle.tsx:25]: we can define `expanded: true` here.

File: plugins/plugin-chart-echarts/src/Timeseries/Regular/controlPanel.tsx [hunk 11]
```diff
@@ -89,6 +89,7 @@ const config: ControlPanelConfig = {
     sections.advancedAnalyticsControls,
     sections.annotationsAndLayersControls,
     sections.forecastIntervalControls,
+    { ...sections.titleControls, expanded: true },
```
Comment (zhaoyongjie) [inline plugins/plugin-chart-echarts/src/Timeseries/Regular/controlPanel.tsx:92]: avoid clone titleControls when defined `expanded: true` in section.
```
sections.titleControls,
```

File: packages/superset-ui-chart-controls/src/sections/chartTitle.tsx [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        },
+      },
+    ],
+    [
+      {
+        name: 'y_axis_title_margin',
+        config: {
+          type: 'SelectControl',
+          freeForm: true,
+          clearable: false,
```
Comment (zhaoyongjie) [inline packages/superset-ui-chart-controls/src/sections/chartTitle.tsx:75]: We should enable user to clean up margin value(xaxis and yaxis both)
image

File: packages/superset-ui-chart-controls/src/utils/titleOptions.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
+export const TITLE_MARGIN_OPTIONS: [number, number][] = [
```
Comment (zhaoyongjie) [inline packages/superset-ui-chart-controls/src/utils/titleOptions.ts:1]: These constants, if used only in `chartTitle.tsx` component, suggest defining within the component.

File: plugins/plugin-chart-echarts/src/Timeseries/types.ts [hunk 14]
```diff
@@ -77,6 +77,10 @@ export type EchartsTimeseriesFormData = QueryFormData & {
   groupby: string[];
   showValue: boolean;
   onlyTotal: boolean;
+  xAxisTitle: string;
```
Comment (zhaoyongjie) [inline plugins/plugin-chart-echarts/src/Timeseries/types.ts:80]: add a `EchartsTitleFormData` type define, then union these types.

**Review by zhaoyongjie — CHANGES_REQUESTED** [review]

File: packages/superset-ui-chart-controls/src/sections/chartTitle.tsx (hunk 8 above)
Comment (MaxHuiYYDS) [reply packages/superset-ui-chart-controls/src/sections/chartTitle.tsx]: Done

File: packages/superset-ui-chart-controls/src/shared-controls/index.tsx (hunk 9 above)
Comment (MaxHuiYYDS) [reply packages/superset-ui-chart-controls/src/shared-controls/index.tsx]: Done

File: packages/superset-ui-chart-controls/src/sections/chartTitle.tsx (hunk 10 above)
Comment (MaxHuiYYDS) [reply packages/superset-ui-chart-controls/src/sections/chartTitle.tsx:26]: Done

File: plugins/plugin-chart-echarts/src/Timeseries/Regular/controlPanel.tsx (hunk 11 above)
Comment (MaxHuiYYDS) [reply plugins/plugin-chart-echarts/src/Timeseries/Regular/controlPanel.tsx]: Done

File: packages/superset-ui-chart-controls/src/sections/chartTitle.tsx (hunk 12 above)
Comment (MaxHuiYYDS) [reply packages/superset-ui-chart-controls/src/sections/chartTitle.tsx]: Done

File: packages/superset-ui-chart-controls/src/utils/titleOptions.ts (hunk 13 above)
Comment (MaxHuiYYDS) [reply packages/superset-ui-chart-controls/src/utils/titleOptions.ts]: Done

File: plugins/plugin-chart-echarts/src/Timeseries/types.ts (hunk 14 above)
Comment (MaxHuiYYDS) [reply plugins/plugin-chart-echarts/src/Timeseries/types.ts]: Done

**Review by zhaoyongjie — APPROVED** [review]
Thanks for your patience. LGTM.
